File: py-dfe/py_dfe/soap/envelope.py
"""SOAP 1.2 envelope builder for SEFAZ web services."""

from __future__ import annotations

import base64
import gzip as gz

from lxml import etree

from py_dfe.constants.endpoints import get_authorizer
from py_dfe.constants.enums import (
    DOC_NAMESPACE,
    SCHEMA_VERSION,
    SOAP_ELEMENTS,
    SOAP_HEADER_VERSION_OVERRIDE,
    SOAP_WRAPPED_BODY_OVERRIDES,
    SOAP_WRAPPED_BODY_SERVICES,
    UF_IBGE,
    WSDL_OPERATION_BY_DOCTYPE,
    WSDL_OPERATION_OVERRIDES,
    WSDL_SERVICE_BY_DOCTYPE,
)
from py_dfe.constants.types import XMLElement
from py_dfe.exceptions import SOAPError
from py_dfe.xmlops.builder import create_element, create_subelement

_SOAP12_NS = "http://www.w3.org/2003/05/soap-envelope"
_XSI_NS = "http://www.w3.org/2001/XMLSchema-instance"
_XSD_NS = "http://www.w3.org/2001/XMLSchema"


def _resolve_operation(doc_type: str, uf: str, service: str) -> str:
    """Return the WSDL operation name for the given combination.

    Checks per-authorizer overrides first, then falls back to the global default.
    """
    authorizer = get_authorizer(doc_type, uf)
    override = WSDL_OPERATION_OVERRIDES.get((authorizer, doc_type, service))
    if override:
        return override
    return WSDL_OPERATION_BY_DOCTYPE[doc_type].get(service, "")


class SOAPEnvelopeBuilder:
    """Builder for SEFAZ SOAP 1.2 envelopes.

    Usage
    -----
    >>> payload_bytes = b'<test></test>'
    >>> builder = SOAPEnvelopeBuilder("nfe", "SP", "NFeAutorizacao")
    >>> envelope_bytes = builder.build(payload_bytes)
    """

    def __init__(self, doc_type: str, uf: str, service: str) -> None:
        self._doc_type = doc_type
        self._uf = uf
        self._service = service

        wsdl_name = WSDL_SERVICE_BY_DOCTYPE[doc_type].get(service)
        if wsdl_name is None:
            raise SOAPError(f"Unknown service {service!r} for doc_type {doc_type!r}")

        base_wsdl = f"{DOC_NAMESPACE[doc_type]}/wsdl"
        self._wsdl_ns = f"{base_wsdl}/{wsdl_name}"
        self._authorizer = get_authorizer(doc_type, uf)
        self._operation = _resolve_operation(doc_type, uf, service)
        self._version = SOAP_HEADER_VERSION_OVERRIDE.get(service, SCHEMA_VERSION[doc_type])
        self._cuf = str(UF_IBGE.get(uf, 0))
        elems = SOAP_ELEMENTS[doc_type]
        self._header_elem = elems["header"]
        self._body_elem = elems["body"]

    def build(self, payload_xml: bytes, gzip=False, include_header=False) -> bytes:
        """Wrap *payload_xml* in a SOAP 1.2 envelope and return as bytes."""
        try:
            envelope = self._make_envelope(payload_xml, gzip, include_header)
            return etree.tostring(envelope, xml_declaration=True, encoding="UTF-8")
        except SOAPError:
            raise
        except Exception as exc:
            raise SOAPError(f"Failed to build SOAP envelope: {exc}") from exc

    @property
    def content_type(self) -> str:
        """Content-Type header value including the SOAP action."""
        return (
            f"application/soap+xml; charset=utf-8; "
            f'action="{self._wsdl_ns}/{self._operation}"'
        )

    def _make_envelope(
        self,
        payload_xml: bytes,
        gzip: bool = False,
        include_header: bool = False,
    ) -> XMLElement:
        nsmap = {
            "soap12": _SOAP12_NS,
            "xsi": _XSI_NS,
            "xsd": _XSD_NS,
        }
        envelope = create_element(f"{{{_SOAP12_NS}}}Envelope", nsmap=nsmap)
        ns = self._wsdl_ns
        if include_header:
            header = create_subelement(envelope, f"{{{_SOAP12_NS}}}Header")
            cab = create_subelement(
                header, f"{{{ns}}}{self._header_elem}", nsmap={None: ns}
            )
            create_subelement(cab, f"{{{ns}}}cUF").text = self._cuf
            create_subelement(cab, f"{{{ns}}}versaoDados").text = self._version

        body = create_subelement(envelope, f"{{{_SOAP12_NS}}}Body")
        if self._service in SOAP_WRAPPED_BODY_SERVICES or (self._authorizer, self._service) in SOAP_WRAPPED_BODY_OVERRIDES:
            wrapper = create_subelement(
                body, f"{{{ns}}}{self._operation}", nsmap={None: ns}
            )
            data = create_subelement(wrapper, f"{{{ns}}}{self._body_elem}")
        else:
            data = create_subelement(
                body, f"{{{ns}}}{self._body_elem}", nsmap={None: ns}
            )

        if gzip:
            data.text = base64.b64encode(gz.compress(payload_xml)).decode("utf-8")
        else:
            payload_root = etree.fromstring(payload_xml)
            data.append(payload_root)

        return envelope


def extract_body(soap_response: bytes) -> bytes:
    """Extract the first child element of the SOAP Body and return as bytes."""
    try:
        root = etree.fromstring(soap_response)
        body = root.find(f"{{{_SOAP12_NS}}}Body")
        if body is None:
            raise SOAPError("SOAP Body element not found in response")
        children = list(body)
        if not children:
            raise SOAPError("SOAP Body is empty")
        return etree.tostring(children[0])
    except SOAPError:
        raise
    except Exception as exc:
        raise SOAPError(f"Failed to extract SOAP body: {exc}") from exc

File: go-dfe/internal/soap/envelope.go
// Package soap ports py-dfe's SOAP 1.2 envelope handling
// (py-dfe/py_dfe/soap/envelope.py) to Go: wrapping a signed/unsigned fiscal
// XML body for a specific SEFAZ service call (Build), and unwrapping SEFAZ's
// SOAP response to get at the inner result XML (ParseResult).
//
// All service/UF-specific lookups (header/body/result element names, the
// wrapped-body special cases, WSDL operation names, header version override)
// come from the constants package — this file has no lookup tables of its own.
package soap

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
	"encoding/xml"
	"fmt"
	"io"
	"strconv"

	"gopkg.aoctech.app/dfe/go-dfe/internal/constants"
)

const soap12NS = "http://www.w3.org/2003/05/soap-envelope"

// Builder builds SOAP 1.2 envelopes for one (docType, uf, service) call,
// mirroring py-dfe's SOAPEnvelopeBuilder. All lookups happen once in
// NewBuilder so Build/ContentType are pure formatting.
type Builder struct {
	wsdlNS    string
	operation string
	version   string
	cUF       string
	wrapped   bool
	headerEl  string
	bodyEl    string
}

// NewBuilder resolves the WSDL namespace/operation/element names for
// (docType, uf, service). Returns an error if service is unknown for docType,
// mirroring py-dfe's SOAPError("Unknown service ...").
func NewBuilder(docType, uf, service string) (*Builder, error) {
	wsdlName, ok := constants.WSDLServiceByDocType[docType][service]
	if !ok {
		return nil, fmt.Errorf("soap: unknown service %q for doc_type %q", service, docType)
	}

	operation, err := constants.WSDLOperation(uf, docType, service)
	if err != nil {
		return nil, fmt.Errorf("soap: %w", err)
	}

	elems, ok := constants.SOAPElements[docType]
	if !ok {
		return nil, fmt.Errorf("soap: no SOAP element names for doc_type %q", docType)
	}

	version := constants.SchemaVersion[docType]
	if v, ok := constants.SOAPHeaderVersionOverride[service]; ok {
		version = v
	}

	wrapped := constants.SOAPWrappedBodyServices[service] || constants.IsSOAPWrappedBodyOverride(uf, service)

	return &Builder{
		wsdlNS:    fmt.Sprintf("%s/wsdl/%s", constants.DocNamespace[docType], wsdlName),
		operation: operation,
		version:   version,
		cUF:       strconv.Itoa(constants.UFIBGE[uf]), // 0 for unknown uf, matching py-dfe's UF_IBGE.get(uf, 0)
		wrapped:   wrapped,
		headerEl:  elems.Header,
		bodyEl:    elems.Body,
	}, nil
}

// ContentType returns the Content-Type header value including the SOAP
// action, matching py-dfe's SOAPEnvelopeBuilder.content_type.
func (b *Builder) ContentType() string {
	return fmt.Sprintf(`application/soap+xml; charset=utf-8; action="%s/%s"`, b.wsdlNS, b.operation)
}

// Build wraps payloadXML (already-serialized fiscal XML, e.g. the output of
// the fiscal XML builder/signer) in a SOAP 1.2 envelope ready to POST to
// SEFAZ. If gzipPayload is true, payloadXML is gzip-compressed and
// base64-encoded into the body element instead of embedded as XML (some
// services require this). If includeHeader is true, a SOAP Header carrying
// cUF/versaoDados is added.
func (b *Builder) Build(payloadXML []byte, gzipPayload, includeHeader bool) ([]byte, error) {
	var buf bytes.Buffer
	buf.WriteString(xml.Header)
	buf.WriteString(`<soap12:Envelope xmlns:soap12="` + soap12NS + `" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">`)

	if includeHeader {
		buf.WriteString(`<soap12:Header>`)
		fmt.Fprintf(&buf, `<%s xmlns="%s"><cUF>%s</cUF><versaoDados>%s</versaoDados></%s>`,
			b.headerEl, b.wsdlNS, b.cUF, b.version, b.headerEl)
		buf.WriteString(`</soap12:Header>`)
	}

	buf.WriteString(`<soap12:Body>`)
	if b.wrapped {
		fmt.Fprintf(&buf, `<%s xmlns="%s"><%s>`, b.operation, b.wsdlNS, b.bodyEl)
	} else {
		fmt.Fprintf(&buf, `<%s xmlns="%s">`, b.bodyEl, b.wsdlNS)
	}

	if gzipPayload {
		var gz bytes.Buffer
		zw := gzip.NewWriter(&gz)
		if _, err := zw.Write(payloadXML); err != nil {
			return nil, fmt.Errorf("soap: gzip payload: %w", err)
		}
		if err := zw.Close(); err != nil {
			return nil, fmt.Errorf("soap: gzip payload: %w", err)
		}
		buf.WriteString(base64.StdEncoding.EncodeToString(gz.Bytes()))
	} else {
		buf.Write(payloadXML)
	}

	if b.wrapped {
		fmt.Fprintf(&buf, `</%s></%s>`, b.bodyEl, b.operation)
	} else {
		fmt.Fprintf(&buf, `</%s>`, b.bodyEl)
	}
	buf.WriteString(`</soap12:Body></soap12:Envelope>`)

	return buf.Bytes(), nil
}

// ParseResult unwraps raw SOAP 1.2 response bytes down to the inner result
// XML: it locates the SOAP Body, then searches it (at any depth, since
// operation-result wrappers vary) for the element named after docType's
// Result element (constants.SOAPElements[docType].Result, e.g.
// "nfeResultMsg") and returns that element's raw bytes.
//
// If no such element is found, it falls back to the Body's first child
// element, mirroring py-dfe's extract_body (py-dfe/py_dfe/soap/envelope.py),
// which does not check element names at all.
func ParseResult(docType string, soapResponse []byte) ([]byte, error) {
	elems, ok := constants.SOAPElements[docType]
	if !ok {
		return nil, fmt.Errorf("soap: no SOAP element names for doc_type %q", docType)
	}

	body, found, err := findElement(soapResponse, soap12NS, "Body")
	if err != nil {
		return nil, err
	}
	if !found {
		return nil, fmt.Errorf("soap: SOAP Body element not found in response")
	}

	if result, found, err := findElement(body, "", elems.Result); err != nil {
		return nil, err
	} else if found {
		return result, nil
	}

	return firstChild(body)
}

// findElement does a document-order search (which, since xml.Decoder emits
// tokens depth-first, is effectively a DFS over the whole tree) for the first
// element whose local name is local and — if ns is non-empty — whose
// namespace is ns. It returns the raw bytes spanning that element's start and
// end tags.
func findElement(data []byte, ns, local string) ([]byte, bool, error) {
	dec := xml.NewDecoder(bytes.NewReader(data))
	for {
		start := dec.InputOffset()
		tok, err := dec.Token()
		if err == io.EOF {
			return nil, false, nil
		}
		if err != nil {
			return nil, false, fmt.Errorf("soap: parse xml: %w", err)
		}
		se, ok := tok.(xml.StartElement)
		if !ok {
			continue
		}
		if se.Name.Local != local || (ns != "" && se.Name.Space != ns) {
			continue
		}
		if err := dec.Skip(); err != nil {
			return nil, false, fmt.Errorf("soap: parse xml: %w", err)
		}
		return data[start:dec.InputOffset()], true, nil
	}
}

// firstChild returns the raw bytes of the first child element of elementXML
// (which must be a single well-formed element, e.g. a SOAP Body), i.e. the
// element one level down from elementXML's own root tag.
func firstChild(elementXML []byte) ([]byte, error) {
	dec := xml.NewDecoder(bytes.NewReader(elementXML))
	depth := 0
	for {
		start := dec.InputOffset()
		tok, err := dec.Token()
		if err == io.EOF {
			return nil, fmt.Errorf("soap: SOAP Body is empty")
		}
		if err != nil {
			return nil, fmt.Errorf("soap: parse xml: %w", err)
		}
		if _, ok := tok.(xml.StartElement); ok {
			depth++
			if depth == 2 {
				if err := dec.Skip(); err != nil {
					return nil, fmt.Errorf("soap: parse xml: %w", err)
				}
				return elementXML[start:dec.InputOffset()], nil
			}
		}
	}
}
